employee/employee.c

```diff
@@ -13,9 +13,14 @@ void change_salary(struct employee *e, double salary)
 	e->salary = salary;
 }
 
+/* FIXME:
+ * It crashes after any `char *`, i.e.
```

Comment:
I can't parse this sentence.

Reply:
My bad, sorry. I rephrase. 
One of the following statements, only ONE: 
```
- printf("%s\n", e->name);
- printf("%s\n", e->surname);
- printt("%s\n", e-->alpha_id);
```
Will not cause any crash. If we put more than one of them, like:
```
printf("%s\n", e->name);
printf("%s\n", e->surname);
```
The Exception will occur.

Reply:
I think what you're saying is that `print_details` crashes on the second `printf` (no matter what that `printf` is, provided it references some part of `e`)?

Reply:
> I think what you're saying is that `print_details` crashes on the second `printf` (no matter what that `printf` is, provided it references some part of `e`)?

Yes. If we try to print more than one element of the `struct`.

Reply:
OK, I have no idea :) @jacobbramley might have a suggestion?

Reply:
This happens after we set the `struct` to *read-only* but we should be able to read, right?

Reply:
Each of those `printf`s dereferences two capabilities (`e` and `name`/`surname`/`alpha_id`). Are you sure it's dereferencing `e` that's failing and not, say `surname`?

Reply:
> Each of those `printf`s dereferences two capabilities (`e` and `name`/`surname`/`alpha_id`). Are you sure it's dereferencing `e` that's failing and not, say `surname`?

But they all have the permissions. Or am I missing something obvious?

Reply:
> Each of those `printf`s dereferences two capabilities (`e` and `name`/`surname`/`alpha_id`). Are you sure it's dereferencing `e` that's failing and not, say `surname`?

As you can see it is failing in `strlen.c` so, yes, it is related to one of those `char *`.

Reply:
I think what Jacob is suggesting is: can you cause this crash without `printf` involved? Since you think `strlen` is the culprit, why not make a test that *only* uses `strlen` and that might help you narrow the problem down.

Reply:
> I think what Jacob is suggesting is: can you cause this crash without `printf` involved? Since you think `strlen` is the culprit, why not make a test that _only_ uses `strlen` and that might help you narrow the problem down.

The following statements do not cause any crash:
```
strlen(ro_employee->alpha_id);
strlen(ro_employee->name);
strlen(ro_employee->surname);
```

Reply:
Can you paste a `gdb` backtrace from the crash or is that not possible?

Reply:
> Can you paste a `gdb` backtrace from the crash or is that not possible?

```
Core was generated by `./read_only'.
Program terminated with signal SIGPROT, CHERI protection violation
Capability sealed fault caused by register ca2.
#0  strlen (str=0x40278e90 <printf+64> [rxR,0x40140000-0x407aa000] (sentry) "\217 \001\001[\021\001\002[\200\300\376[\021\001\376#H\021") at /home/cheriworker/cheri/cheribsd/lib/libc/string/strlen.c:153
```

```
(gdb) info register ca2
ca2            0xf11720000babc5060000000040278e90       0x40278e90 <printf+64> [rxR,0x40140000-0x407aa000] (sentry)
```

Reply:
We need the full backtrace: try `bt` at the gdb command line.

Reply:
```
#0  strlen (str=0x40278e90 <printf+64> [rxR,0x40140000-0x407aa000] (sentry) "\217 \001\001[\021\001\002[\200\300\376[\021\001\376#H\021") at /home/cheriworker/cheri/cheribsd/lib/libc/string/strlen.c:153
#1  0x0000000040280576 in __vfprintf (fp=<optimized out>, locale=<optimized out>, fmt0=<optimized out>, ap=<optimized out>) at /home/cheriworker/cheri/cheribsd/lib/libc/stdio/vfprintf.c:912
#2  0x000000004027eb04 in vfprintf_l (fp=0x403541f0 [rwRW,0x40354020-0x40354590], locale=0x40353620 <__xlocale_global_locale> [rwRW,0x40353620-0x40353780],
    fmt0=0x10075b [rR,0x10075b-0x10076f] "Employee Name: %s \n", ap=0x3fffdffe40 [rwRW,0x3fbfe00000-0x3fffe00000]) at /home/cheriworker/cheri/cheribsd/lib/libc/stdio/vfprintf.c:301
#3  0x0000000040278e90 in printf (fmt=0x40278e90 <printf+64> [rxR,0x40140000-0x407aa000] (sentry) "\217 \001\001[\021\001\002[\200\300\376[\021\001\376#H\021")
    at /home/cheriworker/cheri/cheribsd/lib/libc/stdio/printf.c:57
#4  0x0000000000102808 in print_details (e=0x3fffdffe00 [rR,0x3fffdffe00-0x3fffdffe40]) at employee.c:24
#5  0x00000000001025ea in main () at read_only.c:26
```

Reply:
OK, that doesn't help me, but it might help you and @jacobbramley :)

Reply:
> OK, that doesn't help me, but it might help you and @jacobbramley :)

Can you please check if there is some obvious mistake with the `struct` not related to `CHERI` that I may have done. I don't know, definition, declaration, something I am missing...